src/backend/distributed/connection/connection_management.c

```diff
@@ -527,19 +535,207 @@ FindAvailableConnection(dlist_head *connections, uint32 flags)
 
 
 /*
- * ErrorIfMultipleMetadataConnectionExists throws an error if the
- * input connection dlist contains more than one metadata connections.
+ * EnsureTxStateAndGetHealthyConnections is a helper function that goes
+ * over the input connections, and:
+ *  - Errors for the connections whose remote transaction is broken
+ *    when checkTransactionState=true
+ *  - Skips connections that are not healthy, marks those to be closed
+ *    at the end of the transaction such that they do not linger around
+ *
+ * And, finally returns the list (unless already errored out) of connections
+ * that can are healthy and can be reused.
  */
-static void
-ErrorIfMultipleMetadataConnectionExists(dlist_head *connections)
+static List *
+EnsureTxStateAndGetHealthyConnections(dlist_head *connections,
+									  bool checkTransactionState)
 {
-	bool foundMetadataConnection = false;
+	List *healthyConnections = NIL;
+
 	dlist_iter iter;
 	dlist_foreach(iter, connections)
 	{
 		MultiConnection *connection =
 			dlist_container(MultiConnection, connectionNode, iter.cur);
 
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+		bool inRemoteTrasaction =
+			transaction->transactionState != REMOTE_TRANS_NOT_STARTED;
+
+		if (checkTransactionState && inRemoteTrasaction)
+		{
+			PGTransactionStatusType status = PQtransactionStatus(connection->pgConn);
+
+			/* if the connection is in a bad state, so is the transaction's state */
+			if (status == PQTRANS_INERROR || status == PQTRANS_UNKNOWN)
+			{
+				transaction->transactionFailed = true;
+			}
+
+			if (transaction->transactionFailed &&
+				(transaction->transactionCritical ||
+				 !dlist_is_empty(&connection->referencedPlacements)))
```

Comment:
could use some comments, especially on why we have the referencedPlacements check

Reply:
Added some comments, hope that clarifies the goal for the readers.

We normally don't really check `connection->referencedPlacements`, but seems right thing to do. If you think otherwise, I'm ok to remove that check. 

Btw, I'm not even sure if we can `transaction->transactionCritical` without referenced placements, but still seems more robust to have that